programs/cleos/do_http_post_libcurl.cpp

```diff
@@ -0,0 +1,142 @@
+         case CURLINFO_HEADER_IN: text = "<= Recv header"; break;
+         case CURLINFO_DATA_IN: text = "<= Recv data"; break;
+         case CURLINFO_SSL_DATA_IN: text = "<= Recv SSL data"; break;
+      }
+
+      dump(text, stderr, (unsigned char*)data, size);
+      return 0;
+   }
+
+   struct curl_handle : std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> {
+      curl_handle(int) : std::unique_ptr<CURL, decltype(&curl_easy_cleanup)>( curl_easy_init(), &curl_easy_cleanup) {}
+   };
+
+
+   std::tuple<unsigned int, std::string> do_http_post(const std::string& base_uri, const std::string& path,
+                                                      const std::vector<std::string>& headers,
+                                                      const std::string& postjson, bool verify_cert, bool verbose,
+                                                      bool trace) {
+
+      static bool initialized = false;
+
+      if (!initialized) {
+         auto res = curl_global_init(CURL_GLOBAL_DEFAULT);
+         EOS_ASSERT(res == CURLE_OK, chain::http_exception, curl_easy_strerror(res));
+         initialized = true;
+      }
+
+      static std::map<std::string, curl_handle> handles;
+      auto [itr, _] = handles.emplace(base_uri, 0);
+      auto curl = itr->second.get();
+      EOS_ASSERT(curl != 0, chain::http_exception, "curl_easy_init failed");
+
+      std::string uri;
+
+      const char* unix_socket_prefix     = "unix://";
+      const int   unix_socket_prefix_len = strlen(unix_socket_prefix);
+
+      if (strncmp(unix_socket_prefix, base_uri.c_str(), unix_socket_prefix_len) == 0) {
+         curl_easy_setopt(curl, CURLOPT_UNIX_SOCKET_PATH, base_uri.c_str() + unix_socket_prefix_len);
+         uri = "http://localhost" + path;
+      } else {
+         uri = base_uri + path;
+      }
+
+      curl_easy_setopt(curl, CURLOPT_URL, uri.c_str());
+      curl_easy_setopt(curl, CURLOPT_POST, 1L);
+      curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, (long)CURL_HTTP_VERSION_1_1);
+      curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, postjson.size());
+      curl_easy_setopt(curl, CURLOPT_POSTFIELDS, postjson.c_str());
+
+      if (trace) {
+         curl_easy_setopt(curl, CURLOPT_DEBUGFUNCTION, my_trace);
+         curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
+      }
+
+      if (verbose || trace)
+         curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+
+      struct curl_slist* list = NULL;
+
+      for (auto& h : headers) list = curl_slist_append(list, h.c_str());
+
+      list = curl_slist_append(list, "Expect:");
+      list = curl_slist_append(list, "Content-Type: application/json");
+
+      curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list);
+      curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
+
+      std::string response;
+
+      curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void*)&response);
+      if (verify_cert)
+         curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1L);
```

Comment:
The documentation states the default value for `CURLOPT_SSL_VERIFYPEER` is 1, so this check's logic needs to be inverted.